packages/ckeditor5-markdown-gfm/src/pastefrommarkdownexperimental.ts

```diff
@@ -0,0 +1,195 @@
+
+		const clipboardPipeline: ClipboardPipeline = editor.plugins.get( 'ClipboardPipeline' );
+
+		let shiftPressed = false;
+
+		this.listenTo<ViewDocumentKeyDownEvent>( viewDocument, 'keydown', ( evt, data ) => {
+			shiftPressed = data.shiftKey;
+		} );
+
+		this.listenTo<ClipboardInputTransformationEvent>( clipboardPipeline, 'inputTransformation', ( evt, data ) => {
+			if ( shiftPressed ) {
+				return;
+			}
+
+			const dataAsTextHtml = data.dataTransfer.getData( 'text/html' );
+
+			if ( !dataAsTextHtml ) {
+				const dataAsTextPlain = data.dataTransfer.getData( 'text/plain' );
+				data.content = this._gfmDataProcessor.toView( dataAsTextPlain );
+
+				return;
+			}
+
+			const markdownFromHtml = this._parseMarkdownFromHtml( dataAsTextHtml );
+
+			if ( markdownFromHtml ) {
+				data.content = this._gfmDataProcessor.toView( markdownFromHtml );
+			}
+		} );
+	}
+
+	/**
+	 * Determines if code copied from a website in `text/html` type can be parsed as markdown.
+	 * It removes any OS specific HTML tags e.g. <meta> on Mac OS and <!--StartFragment--> on Windows.
+	 * Then removes a single wrapper HTML tag or wrappers for sibling tags, and if there are no more tags left,
+	 * returns the remaining text. Returns null, if there are any remaining HTML tags detected.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _parseMarkdownFromHtml( htmlString: string ): string | null {
+		const withoutOsSpecificTags = this._removeOsSpecificTags( htmlString );
+
+		const withoutWrapperTag = this._isHtmlList( withoutOsSpecificTags ) ?
+			this._removeListWrapperTagsAndBrs( withoutOsSpecificTags ) :
+			this._removeWrapperTag( withoutOsSpecificTags );
+
+		if ( this._containsAnyRemainingHtmlTags( withoutWrapperTag ) ) {
+			return null;
+		}
+
+		return this._replaceHtmlReservedEntitiesWithCharacters( withoutWrapperTag );
+	}
+
+	/**
+	 * Removes OS specific tags.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _removeOsSpecificTags( htmlString: string ): string {
+		// Removing <meta> tag present on Mac.
+		const withoutMetaTag = htmlString.replace( /^<meta\b[^>]*>/, '' ).trim();
+		// Removing <html> tag present on Windows.
+		const withoutHtmlTag = withoutMetaTag.replace( /^<html>/, '' ).replace( /<\/html>$/, '' ).trim();
+		// Removing <body> tag present on Windows.
+		const withoutBodyTag = withoutHtmlTag.replace( /^<body>/, '' ).replace( /<\/body>$/, '' ).trim();
+
+		// Removing <!--StartFragment--> tag present on Windows.
+		return withoutBodyTag.replace( /^<!--StartFragment-->/, '' ).replace( /<!--EndFragment-->$/, '' ).trim();
+	}
+
+	/**
+	 * Removes a single HTML wrapper tag from string.
+	 *
+	 * @private
+	 * @param htmlString Clipboard content without any OS specific tags.
+	 */
+	private _removeWrapperTag( htmlString: string ): string {
+		return htmlString.replace( /^<(\w+)\b[^>]*>\s*([\s\S]*?)\s*<\/\1>/, '$2' ).trim();
```

Comment:
Can this lead to content loss? Maybe this should be more precise?

Reply:
I removed the complicated regex, because this logic can be handled by this function: `_removeFirstLevelWrapperTagsAndBrs()`.